packages/data-service/src/graphqlObservable.js

```diff
@@ -2,100 +2,149 @@ import { Observable } from "rxjs/Observable";
-    }, Observable.of({}));
+function resolve(types, definition, context, parent) {
+  if (definition.kind === "OperationDefinition") {
+    return resolveOperation(types, definition, context);
   }
 
-  // Node Field
   if (definition.kind === "Field" && definition.selectionSet !== undefined) {
-    const args = definition.arguments
-      .map(arg => {
-        return arg.value.kind === "Variable"
-          ? { [arg.name.value]: context[arg.value.name.value] }
-          : { [arg.name.value]: arg.value.value };
-      })
-      .reduce(Object.assign, {});
-
-    const resolvedObservable = typeMap[definition.name.value].resolve(
-      parent,
-      args,
-      context,
-      null // that would be the info
-    );
+    return resolveNode(types, definition, context, parent);
+  }
 
-    return resolvedObservable.map(emittedResults => {
-      if (!emittedResults.map) {
-        return emittedResults;
-      }
+  if (definition.kind === "Field") {
+    return resolveLeaf(types, definition, context, parent);
+  }
 
-      return emittedResults.map(result => {
-        return definition.selectionSet.selections.reduce((acc, sel) => {
-          acc[sel.name.value] = resolveStep(typeMap, sel, context, result);
+  return throwObservable(`kind not supported "${definition.kind}".`);
+}
 
-          return acc;
-        }, {});
-      });
-    });
-  }
+function resolveOperation(types, definition, context) {
+  const translateOperation = {
+    query: "Query",
+    mutation: "Mutation"
+  };
 
-  // LeafField
-  if (definition.kind === "Field") {
-    return parent[definition.name.value];
+  const nextTypeMap = types[
+    translateOperation[definition.operation]
+  ].getFields();
+
+  return resolveResult(null, nextTypeMap, definition, context);
+}
+
+function resolveNode(types, definition, context, parent) {
+  const args = buildResolveArgs(definition, context);
+  const resolver = types[definition.name.value];
+
+  if (!resolver) {
+    return throwObservable(`missing resolver for ${definition.name.value}`);
   }
 
-  return Observable.throw(
-    new Error(
-      "graphqlObservable error: kind ${definition.kind} is not supported"
-    )
+  const resolvedObservable = resolver.resolve(
+    parent,
+    args,
+    context,
+    null // that would be the info
   );
-}
 
-// eslint-disable-next-line import/prefer-default-export
-export function graphqlObservable(doc, schema, context) {
-  if (doc.definitions.length === 1) {
-    return resolveStep(schema._typeMap, doc.definitions[0], context, null);
+  if (!resolvedObservable) {
+    return throwObservable("resolver returns empty value");
+  }
+
+  if (!(resolvedObservable instanceof Observable)) {
+    return throwObservable("resolver does not return an observable");
   }
 
-  return doc.definitions.reduce((acc, definition) => {
-    const resolvedObservable = resolveStep(
-      schema._typeMap,
+  return resolvedObservable.concatMap(emitted => {
```

Comment:
wonder if there's a convention we could pick from Rx code to name `emitted` better.